Handle compressed debug sections in ELF files

benesch opened the pull request:
@alexcrichton this will need to wait for the `unaligned` feature to land in an actual release of the object crate, but otherwise it's ready to go from my perspective. Let me know what you think! Don't think there's a good way around the `flate2` dependency (I guess we could depend on mini_oxide directly?). Do you think that'll be a showstopper for inclusion in libstd?

----

ELF files allow debug info sections to be compressed. The libbacktrace
backed supported these compressed sections, but the Gimli backend did
not. This commit adds that support to the Gimli backend.

In my tests these debug info sections do not obey the alignment
requirements that the object crate expects for the gABI compression
header (nor can I find a source documenting any alignment requirements),
so this commit additionally enables the "unaligned" feature in the
upcoming version of the object crate.

There is a bit of unsafe to ensure the lifetime of the decompressed
sections matches the lifetime of the mmap'd file. I don't think there is
a way around this unsafe code, unless we are willing to ditch Gimli's
EndianSlice for an (apparently slower) EndianReader backed by a
Cow<[u8]>.

Fix #342.

philipc commented:
> That alignment is something we could respect with the alloc family of functions, but it's not worth dealing with it if we're using unaligned anyway.

Agree that we don't need to check `ch_addralign`, but it's because `gimli` doesn't care about the alignment, not due to the `object`'s `unaligned` feature. DWARF sections shouldn't require alignment anyway and we don't currently check `sh_addralign` for uncompressed DWARF sections either.

alexcrichton commented:
d'oh, right!

I suppose that because this seems like a general ELF thing that maybe non-dwarf sections may need more alignment, but for our use cases definitely doesn't matter much!

benesch commented:
👍 I've removed the check on `ch_addralign`. I also switched the decompression from flate2 to miniz_oxide directly. I _think_ that's everything! Please double check the unsafe code, of course.

alexcrichton commented:
Looks great to me, thanks so much again for your help on this!

benesch commented:
Absolutely! Thanks very much for the quick review and merge!

It occurs to me that the master branch of backtrace now depends on the `object` crate from Git, rather than crates.io. This is certainly fine with me, but not sure if you meant to merge before getting an official release of `object` with the `unaligned` feature, so wanted to surface!

alexcrichton commented:
Ah yeah I'm not too worried about that.

@philipc would you be ok making a new publish of `object` in the near future?

philipc commented:
Yep I can do that.

philipc commented:
`object` 0.20.0 is published.

alexcrichton commented:
Thanks!

This support is now published with 0.3.49
